src/network.zig

```diff
@@ -1203,44 +1180,482 @@ pub const Protocols = struct {
+						.start = range.start +% @as(SequenceIndex, @intCast(buf.len)),
+						.len = range.len - @as(SequenceIndex, @intCast(buf.len)),
+						.timestamp = range.timestamp,
+						.considerForCongestionControl = range.considerForCongestionControl,
+					});
+					range.len = @intCast(buf.len);
+				}
+
+				self.buffer.getSliceAtOffset(@intCast(range.start -% self.fullyConfirmedIndex), buf[0..@intCast(range.len)]) catch unreachable;
+				range.timestamp = time;
+				byteIndex.* = range.start;
+				self.unconfirmedRanges.add(range) catch unreachable;
+				return @intCast(range.len);
+			}
+
+			if(self.highestSentIndex == self.nextIndex) return null;
+			if(self.highestSentIndex +% @as(i32, @intCast(buf.len)) -% self.fullyConfirmedIndex > receiveBufferSize) return null;
+			// Send new packet:
+			const len: SequenceIndex = @min(self.nextIndex -% self.highestSentIndex, @as(i32, @intCast(buf.len)));
+			if(len < buf.len and time -% self.lastUnsentTime < allowedDelay) return null;
+
+			self.buffer.getSliceAtOffset(@intCast(self.highestSentIndex -% self.fullyConfirmedIndex), buf[0..@intCast(len)]) catch unreachable;
+			byteIndex.* = self.highestSentIndex;
+			self.unconfirmedRanges.add(.{
+				.start = self.highestSentIndex,
+				.len = len,
+				.timestamp = time,
+				.considerForCongestionControl = considerForCongestionControl,
+			}) catch unreachable;
+			self.highestSentIndex +%= len;
+			return @intCast(len);
+		}
+	};
+
+	const Channel = struct { // MARK: Channel
+		receiveBuffer: ReceiveBuffer,
+		sendBuffer: SendBuffer,
+		allowedDelay: i64,
+		channelId: ChannelId,
+
+		pub fn init(sequenceIndex: SequenceIndex, delay: i64, id: ChannelId) Channel {
+			return .{
+				.receiveBuffer = .init(),
+				.sendBuffer = .init(sequenceIndex),
+				.allowedDelay = delay,
+				.channelId = id,
+			};
+		}
+
+		pub fn deinit(self: *Channel) void {
+			self.receiveBuffer.deinit();
+			self.sendBuffer.deinit();
+		}
+
+		pub fn connect(self: *Channel, remoteStart: SequenceIndex) void {
+			std.debug.assert(self.receiveBuffer.buffer.len == 0);
+			self.receiveBuffer.availablePosition = remoteStart;
+			self.receiveBuffer.currentReadPosition = remoteStart;
+		}
+
+		pub fn receive(self: *Channel, conn: *Connection, start: SequenceIndex, data: []const u8) !ReceiveBuffer.ReceiveStatus {
+			return self.receiveBuffer.receive(conn, start, data);
+		}
+
+		pub fn send(self: *Channel, protocolIndex: u8, data: []const u8, time: i64) !void {
+			return self.sendBuffer.insertMessage(protocolIndex, data, time);
+		}
+
+		pub fn receiveConfirmationAndGetTimestamp(self: *Channel, start: SequenceIndex) ?SendBuffer.ReceiveConfirmationResult {
+			return self.sendBuffer.receiveConfirmationAndGetTimestamp(start);
+		}
+
+		pub fn checkForLosses(self: *Channel, conn: *Connection, time: i64) usize {
+			const retransmissionTimeout: i64 = @intFromFloat(conn.rttEstimate + 3*conn.rttUncertainty + @as(f32, @floatFromInt(self.allowedDelay)) + 10_000);
```

Comment:
Milliseconds? Magic 10k?

Reply:
Yeah, it is kind of magic, and I don't even know why I put it there. All errors should be covered by the other time estimates.